src/analysis/real/polynomial.lean

```diff
@@ -0,0 +1,131 @@
+-/
+
+open polynomial real set
+
+
+lemma exists_forall_ge_of_polynomial_eval (α : ℝ) (f : polynomial ℝ):
+  ∃ M : ℝ, 0 < M ∧ ∀ (y : ℝ), abs (y - α) ≤ 1 → abs (eval y f) ≤ M :=
+begin
+  obtain ⟨x_max, ⟨h_x_max_range, hM⟩⟩ := is_compact.exists_forall_ge (@compact_Icc (α - 1) (α + 1))
+    ⟨α, le_of_lt $ sub_one_lt _, le_of_lt $ lt_add_one _⟩
+    (continuous_abs.comp f.continuous).continuous_on,
+  use [max (abs (f.eval x_max)) 1, lt_of_lt_of_le zero_lt_one (le_max_right _ _)],
+  intros y hy,
+  have hy' : y ∈ Icc (α - 1) (α + 1),
+  { apply mem_Icc.mpr,
+    have h1 := le_abs_self (y - α),
+    have h2 := neg_le_abs_self (y - α),
+    split; linarith },
+  exact le_trans (hM y hy') (le_max_left _ _),
+end
+
+lemma non_root_interval_of_polynomial (α : ℝ) (f : polynomial ℝ) (h_f_nonzero : f ≠ 0) :
+  ∃ B : ℝ, 0 < B ∧ ∀ x (hr : abs (α - x) < B) (hn : x ≠ α), f.eval x ≠ 0 :=
+begin
+  set f_roots := f.roots.to_finset.erase α,
+  set distances := insert (1 : ℝ) (f_roots.image (λ x, abs (α - x))),
+  have h_nonempty : distances.nonempty := ⟨1, finset.mem_insert_self _ _⟩,
+  set B := distances.min' h_nonempty with hB,
+  have h_allpos : ∀ x : ℝ, x ∈ distances → 0 < x,
+  { intros x hx, rw [finset.mem_insert, finset.mem_image] at hx,
+    rcases hx with rfl | ⟨α₀, ⟨h, rfl⟩⟩,
+    { exact zero_lt_one },
+    { rw [finset.mem_erase] at h,
+      rw [abs_pos, sub_ne_zero], exact h.1.symm }},
+  use [B, (h_allpos B (distances.min'_mem h_nonempty))],
+  intros x hx hxα,
+  have hab₂ : x ∉ f.roots.to_finset,
+  { intro h,
+    have h₁ : x ∈ f_roots, { rw [finset.mem_erase], exact ⟨hxα, h⟩ },
+    have h₂ : abs (α - x) ∈ distances,
+    { rw [finset.mem_insert, finset.mem_image], right, exact ⟨x, ⟨h₁, rfl⟩⟩ },
+    have h₃ := finset.min'_le distances (abs (α - x)) h₂,
+    erw ←hB at h₃, linarith only [lt_of_lt_of_le hx h₃] },
+  rwa [multiset.mem_to_finset, mem_roots h_f_nonzero, is_root.def] at hab₂
+end
+
+lemma non_root_small_interval_of_polynomial (α : ℝ) (f : polynomial ℝ) (h_f_nonzero : f ≠ 0)
+  (M : ℝ) (hM : 0 < M) :
+  ∃ B : ℝ, 0 < B ∧ B ≤ 1 / M ∧ B ≤ 1
+  ∧ ∀ x (hr : abs (α - x) < B) (hn : x ≠ α), f.eval x ≠ 0 :=
+begin
+  obtain ⟨B0, ⟨h_B0_pos, h_B0_root⟩⟩ := non_root_interval_of_polynomial α f h_f_nonzero,
+  have h1M : 0 < 1 / M := one_div_pos.mpr hM,
+  obtain ⟨B1, ⟨hB11, hB12, hB13⟩⟩ : ∃ B1 : ℝ, 0 < B1 ∧ B1 ≤ 1 / M ∧ B1 ≤ B0,
+  { cases le_or_gt (1 / M) B0,
+    { use 1 / M, tauto },
+    { exact ⟨B0, h_B0_pos, le_of_lt h, le_refl B0⟩ }},
+  obtain ⟨B, ⟨hB1, hB2, hB3, hB4⟩⟩ : ∃ B : ℝ, 0 < B ∧ B ≤ 1 / M ∧ B ≤ 1 ∧ B ≤ B0,
+  { cases le_or_gt 1 B1,
+    { use 1, split, norm_num, split, linarith, split, norm_num, linarith },
+    { use B1, exact ⟨hB11, ⟨hB12, ⟨le_of_lt h, hB13⟩⟩⟩ }},
+  refine ⟨B, hB1, hB2, hB3, λ (x : ℝ) (hx : abs (α - x) < B), h_B0_root x _ ⟩,
+  linarith
+end
+
+lemma exists_deriv_eq_slope_of_polynomial_root (α : ℝ) (f : polynomial ℝ) (h_α_root : f.eval α = 0)
```

Comment:
Is this the mean value theorem or am I missing something?

Reply:
Yes, this is almost mean value theorem except that mean value theorem is stated using `deriv` for a general function, here it is more convenient to use `polynomial.derivative`

Reply:
There is `polynomial.has_deriv_at` that compares the formal and analytic derivative. Is it a lot of work using it with the mean value theorem? It seems the more natural proof to me.